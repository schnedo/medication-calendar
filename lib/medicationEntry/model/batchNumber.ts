export default interface BatchNumber {
  readonly number: number;
}

const prefix = "Ch.-B.";
export function format(batchNumber: BatchNumber): string {
  return `${prefix} ${batchNumber.number}`;
}
