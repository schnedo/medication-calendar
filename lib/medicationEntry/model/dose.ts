export default interface Dose {
  readonly amount: number;
}

const unit = "ml";

export function format(dose: Dose): string {
  return `${dose.amount} ${unit}`;
}
