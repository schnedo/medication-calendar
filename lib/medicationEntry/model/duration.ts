export default interface Duration {
  minutes: number;
  hours: number;
}

function formatPart(number: number | undefined) {
  if (number === undefined) {
    return "00";
  }
  const raw = number.toString();
  return raw.length === 1 ? `0${raw}` : raw;
}

export function format(duration: Duration): string {
  return `${formatPart(duration.hours)}:${formatPart(duration.minutes)}`;
}
