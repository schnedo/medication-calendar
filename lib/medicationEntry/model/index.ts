export { default as BatchNumber } from "./batchNumber";
export { default as Dose } from "./dose";
export { default as Duration } from "./Duration";
export { default as Medicament } from "./medicament";
export { default as Medication } from "./medication";
export { default as MedicationEntry } from "./medicationEntry";

export * as Repository from "./repository";
