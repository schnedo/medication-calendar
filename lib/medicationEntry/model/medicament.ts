import BatchNumber from "./batchNumber";

export default interface Medicament {
  readonly name: string;
  readonly batchNumber: BatchNumber;
}
