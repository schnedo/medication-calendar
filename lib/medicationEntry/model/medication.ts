import Dose from "./dose";
import Medicament from "./medicament";
import { v4 } from "uuid";

export default class Medication {
  readonly id: string = v4();

  constructor(readonly dose: Dose, readonly medicament: Medicament) {}
}
