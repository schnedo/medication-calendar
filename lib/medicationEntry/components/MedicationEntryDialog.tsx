import {
  MouseEventHandler,
  ReactElement,
  useEffect,
  useRef,
  useState,
} from "react";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  makeStyles,
  TextField,
} from "@material-ui/core";
import { DatePicker } from "@material-ui/pickers";
import { ModalProps } from "@material-ui/core/Modal";
import { Medication, MedicationEntry } from "../model";
import DurationInput from "./DurationInput";
import MedicationsInput from "./MedicationsInput";

const useStyles = makeStyles((theme) => ({
  dialogContent: {
    display: "flex",
    flexDirection: "column",
    "&>*": {
      marginBottom: theme.spacing(2),
      "&:last-child": {
        marginBottom: theme.spacing(0),
      },
    },
  },
}));

const defaultDate = () => new Date();
const defaultDuration = {
  minutes: 0,
  hours: 0,
};
const defaultBodyMass = "60";
const defaultComments = "";
const defaultMedications: Medication[] = [];

export interface MedicationEntryDialogProps {
  open: boolean;
  medicationEntry?: MedicationEntry;
  onClose?: ModalProps["onClose"];
  onSubmit?: (entry: MedicationEntry) => Promise<void>;
  onAbort?: () => void;
}

// eslint-disable-next-line sonarjs/cognitive-complexity
export default function MedicationEntryDialog({
  open,
  medicationEntry,
  onClose,
  onSubmit,
  onAbort,
}: MedicationEntryDialogProps): ReactElement {
  const idRef = useRef(medicationEntry?.id ?? "");
  const [date, setDate] = useState(medicationEntry?.date ?? defaultDate());
  const [duration, setDuration] = useState(
    medicationEntry?.duration ?? defaultDuration,
  );
  const [bodyMass, setBodyMass] = useState(
    medicationEntry?.bodyMass.amount.toString() ?? defaultBodyMass,
  );
  const [comments, setComments] = useState(
    medicationEntry?.comments ?? defaultComments,
  );
  const [medications, setMedications] = useState<Medication[]>(
    medicationEntry?.medications ?? defaultMedications,
  );

  useEffect(() => {
    if (open) {
      setDate(medicationEntry?.date ?? defaultDate());
      setDuration(medicationEntry?.duration ?? defaultDuration);
      setBodyMass(
        medicationEntry?.bodyMass.amount.toString() ?? defaultBodyMass,
      );
      setComments(medicationEntry?.comments ?? defaultComments);
      setMedications(medicationEntry?.medications ?? defaultMedications);
      idRef.current = medicationEntry?.id ?? "";
    }
  }, [medicationEntry, open]);

  const handleSubmit: MouseEventHandler | undefined = onSubmit
    ? async () => {
        await onSubmit({
          id: idRef.current,
          date,
          duration,
          bodyMass: {
            amount: parseInt(bodyMass),
          },
          comments,
          medications,
        });
      }
    : undefined;

  const { dialogContent } = useStyles();
  return (
    <Dialog open={open} onClose={onClose}>
      <DialogTitle>
        {medicationEntry ? "Eintrag bearbeiten" : "Neuer Eintrag"}
      </DialogTitle>
      <DialogContent className={dialogContent}>
        <TextField
          id={"bodyMassInput"}
          label={"KÃ¶rpergewicht"}
          type={"number"}
          value={bodyMass}
          onChange={(event) => setBodyMass(event.target.value)}
        />
        <DatePicker
          id={"dateInput"}
          label={"Datum"}
          value={date}
          onChange={(value) => setDate(value ?? date)}
        />
        <DurationInput value={duration} onChange={setDuration} />
        <TextField
          id={"commentsInput"}
          label={"Kommentare"}
          multiline
          value={comments}
          onChange={(event) => setComments(event.target.value)}
        />
        <MedicationsInput value={medications} onChange={setMedications} />
      </DialogContent>
      <DialogActions>
        <Button color="primary" onClick={onAbort}>
          Abbrechen
        </Button>
        <Button color="primary" onClick={handleSubmit}>
          Speichern
        </Button>
      </DialogActions>
    </Dialog>
  );
}
