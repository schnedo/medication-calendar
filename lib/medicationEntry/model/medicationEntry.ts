import Medication from "./medication";
import { BodyMass } from "../../contact/model";
import Duration from "./duration";
import { v4 } from "uuid";

export default class MedicationEntry {
  readonly id: string = v4();

  constructor(
    readonly date: Date,
    readonly medications: Medication[],
    readonly duration: Duration,
    readonly comments: string,
    readonly bodyMass: BodyMass,
  ) {}
}
