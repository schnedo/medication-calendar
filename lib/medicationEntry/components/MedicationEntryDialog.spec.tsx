import {
  Matcher,
  render,
  SelectorMatcherOptions,
  waitFor,
} from "@testing-library/react";
import MedicationEntryDialog from "./MedicationEntryDialog";
import { mock } from "@userlike/joke";
import userEvent from "@testing-library/user-event";
import { Duration, Medication, MedicationEntry } from "../model";
import { MuiPickersUtilsProvider } from "@material-ui/pickers";
import DateFnsUtils from "@date-io/date-fns";
import { format } from "date-fns";

const { default: DurationInput } = mock(import("./DurationInput"));
const { default: MedicationsInput } = mock(import("./MedicationsInput"));

function expectFieldsEqualEntry(
  medicationEntry: MedicationEntry,
  query: (
    text: Matcher,
    options?: SelectorMatcherOptions | undefined,
    waitForElementOptions?: unknown,
  ) => HTMLElement,
) {
  // eslint-disable-next-line sonarjs/no-duplicate-string
  expect(query("Körpergewicht")).toHaveDisplayValue(
    medicationEntry.bodyMass.amount.toString(),
  );
  expect(query("Datum")).toHaveDisplayValue(
    format(medicationEntry.date, "MMMM do"),
  );
  expect(DurationInput).toHaveBeenLastCalledWith(
    expect.objectContaining({ value: medicationEntry.duration }),
    {},
  );
  expect(query("Kommentare")).toHaveDisplayValue(medicationEntry.comments);
  expect(MedicationsInput).toHaveBeenLastCalledWith(
    expect.objectContaining({ value: medicationEntry.medications }),
    {},
  );
}

describe("MedicationEntryDialog", () => {
  beforeEach(() => {
    DurationInput.mockReturnValue(<div>mocked Duration Input</div>);
    MedicationsInput.mockReturnValue(<div>mocked MedicationsInput</div>);
  });

  it("should render correctly", async () => {
    expect.hasAssertions();

    jest.setSystemTime(new Date(2021, 1, 13));

    const { container } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={true} />
      </MuiPickersUtilsProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  // waitFor does throw "TypeError: Cannot read property 'createEvent' of null" in CI Pipeline
  // eslint-disable-next-line jest/no-disabled-tests
  it.skip("should call onSubmit when submitting", async () => {
    expect.hasAssertions();

    const onSubmit = jest.fn();

    jest.setSystemTime(new Date(2021, 1, 1));

    const duration: Duration = { minutes: 15, hours: 2 };
    DurationInput.mockImplementation(({ onChange }) => (
      <div data-testid={"durationInput"} onClick={() => onChange(duration)}>
        mocked DurationInput
      </div>
    ));

    const medications: Medication[] = [
      {
        id: "1",
        dose: {
          amount: 200,
        },
        medicament: {
          name: "name",
          batchNumber: {
            number: 12345,
          },
        },
      },
    ];
    MedicationsInput.mockImplementation(({ onChange }) => (
      <div
        data-testid={"medicationsInput"}
        onClick={() => onChange && onChange(medications)}
      >
        mocked MedicationsInput
      </div>
    ));

    const { getByRole, getByLabelText, getByTestId, getByText } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={true} onSubmit={onSubmit} />
      </MuiPickersUtilsProvider>,
    );

    await userEvent.type(
      getByLabelText("Körpergewicht"),
      "{backspace}{backspace}10",
    );
    userEvent.click(getByLabelText("Datum"));
    userEvent.click(getByRole("button", { name: "10" }));
    userEvent.click(getByRole("button", { name: "OK" }));
    userEvent.click(getByTestId("durationInput"));
    await userEvent.type(getByLabelText("Kommentare"), "comments");
    userEvent.click(getByTestId("medicationsInput"));
    userEvent.click(getByText("Speichern"));

    await waitFor(() => {
      expect(onSubmit).toHaveBeenCalledTimes(1);
    });

    const expected: MedicationEntry = {
      bodyMass: {
        amount: 10,
      },
      comments: "comments",
      date: new Date(2021, 1, 10),
      duration: { minutes: 15, hours: 2 },
      id: "",
      medications,
    };
    expect(onSubmit).toHaveBeenLastCalledWith(expected);
  });

  it("should call onAbort when aborting", async () => {
    expect.hasAssertions();

    const onAbort = jest.fn();

    jest.setSystemTime(new Date(2021, 1, 1));

    const { getByRole } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={true} onAbort={onAbort} />
      </MuiPickersUtilsProvider>,
    );

    userEvent.click(getByRole("button", { name: "Abbrechen" }));
    await waitFor(() => {
      expect(onAbort).toHaveBeenCalledTimes(1);
    });
  });

  it("should set values from medicationEntry prop if present", async () => {
    expect.hasAssertions();
    const medicationEntry1: MedicationEntry = {
      bodyMass: { amount: 20 },
      comments: "comments1",
      date: new Date(2021, 1, 2),
      duration: {
        minutes: 15,
        hours: 3,
      },
      id: "1",
      medications: [],
    };
    const medicationEntry2: MedicationEntry = {
      bodyMass: { amount: 25 },
      comments: "comments2",
      date: new Date(2021, 2, 3),
      duration: {
        minutes: 10,
        hours: 2,
      },
      id: "2",
      medications: [],
    };

    const { getByLabelText, rerender } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={true} medicationEntry={medicationEntry1} />
      </MuiPickersUtilsProvider>,
    );

    expectFieldsEqualEntry(medicationEntry1, getByLabelText);

    rerender(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={false} />
      </MuiPickersUtilsProvider>,
    );
    rerender(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={true} medicationEntry={medicationEntry2} />
      </MuiPickersUtilsProvider>,
    );

    expectFieldsEqualEntry(medicationEntry2, getByLabelText);
  });

  it("should reset state when closing and opening again", async () => {
    expect.hasAssertions();

    const date = new Date(2021, 1, 1);
    jest.setSystemTime(date);

    const duration: Duration = { minutes: 15, hours: 2 };
    // eslint-disable-next-line sonarjs/no-identical-functions
    DurationInput.mockImplementation(({ onChange }) => (
      <div data-testid={"durationInput"} onClick={() => onChange(duration)}>
        mocked DurationInput
      </div>
    ));

    const medications: Medication[] = [
      {
        id: "1",
        dose: {
          amount: 200,
        },
        medicament: {
          name: "name",
          batchNumber: {
            number: 12345,
          },
        },
      },
    ];
    // eslint-disable-next-line sonarjs/no-identical-functions
    MedicationsInput.mockImplementation(({ onChange }) => (
      <div
        data-testid={"medicationsInput"}
        onClick={() => onChange && onChange(medications)}
      >
        mocked MedicationsInput
      </div>
    ));

    const { getByRole, getByLabelText, getByTestId, rerender } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={true} />
      </MuiPickersUtilsProvider>,
    );

    const medicationEntry: MedicationEntry = {
      bodyMass: { amount: 10 },
      comments: "comments",
      date: new Date(date.getFullYear(), date.getMonth(), 10),
      duration,
      id: "1",
      medications,
    };

    await userEvent.type(
      getByLabelText("Körpergewicht"),
      `{backspace}{backspace}${medicationEntry.bodyMass.amount}`,
    );
    userEvent.click(getByLabelText("Datum"));
    userEvent.click(
      getByRole("button", { name: `${medicationEntry.date.getDate()}` }),
    );
    userEvent.click(getByRole("button", { name: "OK" }));
    userEvent.click(getByTestId("durationInput"));
    await userEvent.type(
      getByLabelText("Kommentare"),
      medicationEntry.comments,
    );
    userEvent.click(getByTestId("medicationsInput"));

    expectFieldsEqualEntry(medicationEntry, getByLabelText);

    rerender(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={false} />
      </MuiPickersUtilsProvider>,
    );
    rerender(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <MedicationEntryDialog open={true} />
      </MuiPickersUtilsProvider>,
    );

    const expected: MedicationEntry = {
      bodyMass: { amount: 60 },
      comments: "",
      date,
      duration: { minutes: 0, hours: 0 },
      id: "",
      medications: [],
    };

    expectFieldsEqualEntry(expected, getByLabelText);
  });
});
