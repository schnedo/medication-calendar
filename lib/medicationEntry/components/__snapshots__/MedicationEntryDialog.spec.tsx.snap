// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MedicationEntryDialog should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
